src/hotspot/share/gc/z/zDirector.cpp

```diff
@@ -518,7 +517,11 @@ static bool rule_major_allocation_rate(const ZDirectorStats& stats) {
 
   // If the garbage is cheaper to reap in the old generation, then it makes sense
   // to upgrade minor collections to major collections.
-  const bool old_garbage_is_cheaper = current_old_gc_time_per_bytes_freed < current_young_gc_time_per_bytes_freed;
+  bool old_garbage_is_cheaper = true;
+  if (reclaimed_per_old_gc != 0) {
+    const double current_old_gc_time_per_bytes_freed = double(old_gc_time) / double(reclaimed_per_old_gc);
+    old_garbage_is_cheaper = current_old_gc_time_per_bytes_freed < current_young_gc_time_per_bytes_freed;
+  }
```

Comment:
Ending up with `old_garbage_is_cheaper == true` when  `reclaimed_per_old_gc == 0` seems wrong to me. 

Division by 0.0 is weird in C++. Do we even build for systems where it would not be supported. But regardless to me I feel like the change here should be more like:
```diff
-const double current_old_gc_time_per_bytes_freed = double(old_gc_time) / double(reclaimed_per_old_gc);
+ const double current_old_gc_time_per_bytes_freed = reclaimed_per_old_gc == 0 ? std::numeric_limits<double>::infinity : double(old_gc_time) / double(reclaimed_per_old_gc);
```
Which is the behaviour I expect us to currently have, given that `old_gc_time` should be a positive number (`>0.0`).  The `!stats._old_stats._cycle._is_time_trustable` check above should protect against `0.0`.

I expect that this division we see happens when we have run a warmup major collection which did no reclaim any memory.  And this should trigger us to try and promote a minor collection to a major collection.

I am no expert on our supported platforms matrix w.r.t. floating numbers and `std::numeric_limits<T>::has_infinity`.